azure-iot-device/azure/iot/device/provisioning/registration_client_factory.py

```diff
@@ -0,0 +1,40 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+This module provides factory methods for creating clients which would
+communicate with Device Provisioning Service.
+"""
+from .security.sk_security_client import SymmetricKeySecurityClient
+
+from .sk_registration_client import SymmetricKeyRegistrationClient
+from .transport.state_based_mqtt_provider import StateBasedMQTTProvider
+
+
+def create_from_security_client(provisioning_host, security_client, transport_choice):
+    """
+    Creates different types of registration clients which can enable devices to communicate with Device Provisioning
+    Service based on parameters passed.
+    :param provisioning_host: Host running the Device Provisioning Service. Can be found in the Azure portal in the
+    Overview tab as the string Global device endpoint
+    :param security_client: Instance of Security client object which can be either of Symmetric Key, TPM or X.509
+    :param transport_choice: A string representing the transport the user wants
+    :return: A specific registration client based on parameters and validations.
+    """
+    if not provisioning_host:
+        raise ValueError("Provisioning Host must be provided.")
```

Comment:
I'm not sure the value of checking this. I get that at API level we should be more rigid about checking inputs, but this really only checks that the input isn't None, which doesn't even remotely begin to cover all the bad inputs this could be. I think at a certain point with Python, because there are so many edge cases for bad values, we probably shouldn't check what was taken for granted by the contract. In this case, the parameter isn't listed as optional.

Why stop here? Why not also check that security_client and transport_choice aren't None too? Once you begin this level of parameter checking, there's really no end.

Not that any of this matters that much anyway though - this will become internal before long, since we want to get rid of two-step client instantiation.

Reply:
still a valid discussion to have.
one thing i'd ask is if we don't check it, and try to use it, what exception do we get ? if we get something that says "host not found" then great, it's actionable. if it's some obscure unusable thing, then we need to find a way to catch it and throw a helpful error.

---
In reply to: [277773014](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277773014) [](ancestors = 277773014)

Reply:
These are the 3 error that we receive when 
1. No host is given  :- ValueError: Invalid host.
2. Blank host is given socket.gaierror: :- [Errno 11001] getaddrinfo failed
3. Bad value for host is given :- TimeoutError: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond

I will remove this then , but we must discuss where to catch all these error. Mostly it should happen in the provider connect 
There is a easy way to check transport protocol as supplied by user and I think we should since that can be only of 3 possible values (mqtt, amqp, and http).


---
In reply to: [277848292](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277848292) [](ancestors = 277848292,277773014)